File: leetcode/merge-k-sorted-lists.cpp
// https://leetcode.com/problems/merge-k-sorted-lists

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        
        multimap<int, ListNode*> m;
        for (ListNode* p : lists)
            if (p)
                m.emplace(p->val, p);
        
        if (m.empty())
            return nullptr;
        
        
        ListNode* root = begin(m)->second;
        m.erase(begin(m));
        ListNode* p = root->next;
        if (p)
            m.emplace(p->val, p);
        
        // cout << root->val << "\n";
        ListNode* cur = root;
        while (!m.empty()) {
            auto [val, node] = *begin(m);
            // cout << val << "\n";
            cur->next = node;
            cur = cur->next;
            
            m.erase(begin(m));
            ListNode* p = node->next;
            if (p)
                m.emplace(p->val, p);
        }
        
        return root;
    }
};

File: leetcode/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        gen = count()
        h = []
        for head in lists:
            if head:
                heappush(h, (head.val, next(gen), head))

        if not h:
            return None

        _, _, root = heappop(h)
        if root.next:
            heappush(h, (root.next.val, next(gen), root.next))

        cur = root
        while h:
            _, _, node = heappop(h)
            cur.next = node
            cur = cur.next

            if node.next:
                heappush(h, (node.next.val, next(gen), node.next))

        return root